math/optimized_trig.h

```diff
@@ -0,0 +1,233 @@
+/*
+	Copyright 2021 - Kenn Sebesta
+
+	This file is part of the VESC firmware.
+
+	The VESC firmware is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    The VESC firmware is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+    */
+
+#ifndef OPTIMIZED_TRIG_H_
+#define OPTIMIZED_TRIG_H_
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <math.h>
+#include <stdlib.h>
+
+/**
+ * @brief cosine6TaylorApproximation Fixed-point 6th-order taylor approximation which returns cosine(x)
+ * @param AQ     aq * 2^qN
+ * @param BQm   -bq * 2^qN
+ * @param CQ     cq * 2^qN
+ * @param DQm   -dq * 2^(16 + qN - 2*qX)
+ * @param X2    x^2 * 2^(2qX)
+ * @param X2_16 x^2 * 2^16
+ * @return
+ */
+__attribute__((always_inline))
+inline static int32_t cosine6TaylorApproximation(const int32_t AQ, const int32_t BQm, const int32_t CQ, const int32_t DQm, const int32_t X2, const int32_t X2_16) {
+   /*
+    * out = a - bx^2 + cx^4 - dx^6
+    *     = a - x^2*((b - x^2*(c - d*x^2))
+    *     = a - x^2*(((-bm) - x^2*(c - (-dm)*x^2))
+    *     = a + x^2*(bm + x^2*(c + dm*x^2))
+    */
+
+   int32_t output;
+#if __thumb2__  // Only compile this branch
+   // ARM THUMB-2 assembly blob which takes advantage of the SMLAWb operator, which performs a multily and accumulate
+   asm ("SMLAWb %[Y3], %[X2], %[D], %[C]\n"       // (y1 = cq  + dqm*(x)^2)
+        "SMLAWb %[Y3], %[Y3], %[X2_16], %[B]\n"   // (y2 = bqm +  y1*(x16)^2)
+        "SMLAWb %[Y3], %[Y3], %[X2_16], %[A]\n"   // (y3 = aq  +  y2*(x16)^2)
+        : [Y3] "=&r" (output)
+        : [A] "r" (AQ), [B] "r" (BQm), [C] "r" (CQ), [D] "r" (DQm), [X2] "r" (X2), [X2_16] "r" (X2_16)
+        :);
+
+#else
+   int Y3=(((int64_t)X2*DQm)>>16)+CQ; // (y1=c-d*x^2)
+   Y3=(((int64_t)Y3*X2_16)>>16)+BQm; // (y2=-(b-y1*x16^2))
+   Y3=(((int64_t)Y3*X2_16)>>16)+AQ; // (y3=y2*x16^2+a)
+   output = Y3;
+
+#endif
+
+   return output;
+}
+
+
+/**
+ * @brief _SinCos66_FP This function uses a 6th order taylor expansion for atomically calculating
+ *                     sine(x) and cos(x). It uses fixed-point calculations, and thus is sensitive
+ *                     to the scale of the arguments. Please see the below notes.
+ *
+ *                     Coefficients were chosen by setting:
+ *                       *) cos(0) = 1
+ *                       *) cos(pi/2) = 0
+ *                       *) d/dx[cos(x=pi/2)] = -1
+ *                       *) Minimizing the root-mean square error
```

Comment:
It may be better to minimize maximum absolute/relative error instead of RMS (infinite-norm). Mean error is a problematic is there is small part of argument space where error is large. Also both relative and absolute error is important .. 

Look at https://github.com/samhocevar/lolremez, IICR it was quite easy to optimize polynomial coefficients for (floating point) sin/cos approximation

Reply:
I had never seen LolRemez, I love it.

Having looked at the graphs, relative error is acceptably small, like <1 e-4 across much fo the range. The biggest driver of error is the limits of integer precision and dynamic range. When the value of `cos(x)` starts getting very close to 0 the relative error gets really bad because there's only so much you can express with an integer value which itself is close to 0. 

This is why I switch to the approx `cos(x) ~= pi/2-x` for values of `x` close to pi/2. This in theory adds a tiny jump in the graphed curve, however in reality this jump is lost in the noise of cos(x) values which are less than 1e-3.